spec/java/util/ArrayList.lsl

```diff
@@ -0,0 +1,1205 @@
+        action NOT_IMPLEMENTED();
+    }
+
+
+    fun remove (o: Object): boolean
+    {
+        var index = action LIST_FIND(storage, o, 0, length, +1);
+        if (index == -1)
+        {
+            result = false;
+        }
+        else
+        {
+            action LIST_REMOVE(storage, index, 1);
+            result = true;
+        }
+    }
+
+
+    fun clear (): void
+    {
+        action LIST_RESIZE(storage, 0);
+        length = 0;
+        modCount = modCount + 1;
+    }
+
+
+    fun addAll (c: Collection): boolean
+    {
+        result = _addAllElements(length, c);
+    }
+
+
+
+    fun addAll (index: int, c: Collection): boolean
+    {
+        _rangeCheckForAdd(index, length);
+        result = _addAllElements(index, c);
+    }
+
+
+    fun removeAll (c: Collection): boolean
+    {
+        // #problem.
+        //I want to create such action:
+
+        //define action SUBTRACTING_SETS(
+        //        aListSubtracting: array<any>,
+        //        aListDeductible: array<any>
+        //    ): list<any>;
+
+        //But Collection isn't List.
+        //But it has method: toArray();
+        //Maybe we must have two arrays as params ?
+
+        //storage = action SUBTRACTING_SETS(storage, c);
+
+        action NOT_IMPLEMENTED();
+    }
+
+
+    fun retainAll (c: Collection): boolean
+    {
+        // #problem.
+        //I want to create such action:
+
+        //define action INTERSECTION(
+        //        aListSubtracting: array<any>,
+        //        aListDeductible: array<any>
+        //    ): list<any>;
+
+        //But Collection isn't List.
+        //But it has method: toArray();
+        //Maybe we must have two arrays as params ?
+
+        //storage = action INTERSECTION(storage, c);
+
+        action NOT_IMPLEMENTED();
+    }
+
+
+    @private
+    @throws(["java.io.IOException"])
+    fun writeObject (s: ObjectOutputStream): void
+    {
+        var expectedModCount = modCount;
+
+        // #problem
+        //Here is cycle and some operations with ObjectOutputStream.
+        action NOT_IMPLEMENTED();
+
+        if (modCount != expectedModCount)
+        {
+            action THROW_NEW("java.util.ConcurrentModificationException", []);
+        }
+    }
+
+
+    @private
+    @throws(["java.io.IOException", "java.lang.ClassNotFoundException"])
+    fun readObject (s: ObjectInputStream): void
+    {
+        // #problem
+        //Here is cycle and some operations with ObjectInputStream.
+        action NOT_IMPLEMENTED();
+    }
+
+
+    fun listIterator (index: int): ListIterator
+    {
+        _rangeCheckForAdd(index, length);
+
+        result = new ListItr(state=Created,
+            cursor=index,
+            expectedModCount=modCount);
+    }
+
+
+    fun listIterator (): ListIterator
+    {
+        result = new ListItr(state=Created,
+            cursor=0,
+            expectedModCount=modCount);
+    }
+
+
+    fun iterator (): Iterator
+    {
+        result = new Itr(state=Created,
```

Comment:
I don't see why we should use two different classes/automata where one is just an extension of another.
Why not just merge them in one and use only `ListItr`?

Reply:
Fixed